cpu/nrf5x_common/radio/nrfmin/nrfmin.c

```diff
@@ -0,0 +1,549 @@
+} state_t;
+
+
+/**
+ * @brief   Since there can only be 1 nrfmin device, we allocate it right here
+ */
+netdev2_t nrfmin_dev;
+
+/**
+ * @brief   For faster lookup we remember our own 16-bit address
+ */
+static uint16_t my_addr;
+
+/**
+ * @brief   We need to keep track of the radio state in SW (-> PAN ID 20)
+ *
+ * See nRF51822 PAN ID 20: RADIO State Register is not functional.
+ */
+static volatile state_t state = STATE_OFF;
+
+/**
+ * @brief   We also remember the 'long-term' state, so we can resume after TX
+ */
+static volatile state_t target_state = STATE_OFF;
+
+/**
+ * @brief   When sending out data, the data needs to be in one continuous memory
+ *          region. So we need to buffer outgoing data on the driver level.
+ */
+static nrfmin_pkt_t tx_buf;
+
+/**
+ * @brief   As the device is memory mapped, we need some space to save incoming
+ *          data to.
+ *
+ * @todo    Improve the RX buffering to at least use double buffering
+ */
+static nrfmin_pkt_t rx_buf;
+
+/**
+ * @brief   While we listen for incoming data, we lock the RX buffer
+ */
+static volatile uint8_t rx_lock = 0;
+
+/**
+ * @brief   Set radio into idle (DISABLED) state
+ */
+static void go_idle(void)
+{
+    /* set device into basic disabled state */
+    NRF_RADIO->EVENTS_DISABLED = 0;
+    NRF_RADIO->TASKS_DISABLE = 1;
+    while (NRF_RADIO->EVENTS_DISABLED == 0) {}
+    /* also release any existing lock on the RX buffer */
+    rx_lock = 0;
+    state = STATE_IDLE;
+}
+
+/**
+ * @brief   Set radio into the target state as defined by `target_state`
+ *
+ * Trick here is, that the driver can go back to it's previous state after a
+ * send operation, so it can differentiate if the driver was in DISABLED or in
+ * RX mode before the send process had started.
+ */
+static void goto_target_state(void)
+{
+    go_idle();
+
+    if ((target_state == STATE_RX) && (rx_buf.pkt.hdr.len == 0)) {
+        /* set receive buffer and our own address */
+        rx_lock = 1;
+        NRF_RADIO->PACKETPTR = (uint32_t)(&rx_buf);
+        NRF_RADIO->BASE0 = (CONF_ADDR_BASE | my_addr);
+        /* goto RX mode */
+        NRF_RADIO->TASKS_RXEN = 1;
+        state = STATE_RX;
+    }
+
+    if (target_state == STATE_OFF) {
+        NRF_RADIO->POWER = 0;
+        state = STATE_OFF;
+    }
+}
+
+void nrfmin_setup(void)
+{
+    nrfmin_dev.driver = &nrfmin_netdev;
+    nrfmin_dev.event_callback = NULL;
+    nrfmin_dev.context = NULL;
+#ifdef MODULE_NETSTATS_L2
+    memset(&nrfmin_dev.stats, 0, sizeof(netstats_t));;
+#endif
+}
+
+uint16_t nrfmin_get_addr(void)
+{
+    return my_addr;
+}
+
+void nrfmin_get_pseudo_long_addr(uint16_t *addr)
+{
+    for (int i = 0; i < 4; i++) {
+        addr[i] = my_addr;
+    }
+}
+
+void nrfmin_get_iid(uint16_t *iid)
+{
+    iid[0] = 0;
+    iid[1] = 0xff00;
+    iid[2] = 0x00fe;
+    iid[3] = my_addr;
+}
+
+uint16_t nrfmin_get_channel(void)
+{
+    return (uint16_t)(NRF_RADIO->FREQUENCY >> 2);
```

Comment:
Yes: this is wanted: I changed the channel handling from the previous version, now i mapped the possible frequencies to 32 channels with 4MHz spacing...